src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs

```diff
@@ -236,6 +236,122 @@ public static void Main()
+                Assert.True(emitResult.Success);
+                emitResult.Diagnostics.Verify();
+                image1 = output.ToArray();
+            }
+
+            var source2 = sourceTemplate.Replace("CHANGE", change2);
+            Compilation comp2 = CreateCompilationWithMscorlib(Parse(source2),
+                options: TestOptions.DebugDll.WithDeterministic(true), assemblyName: name);
+
+            byte[] image2;
+
+            using (var output = new MemoryStream())
+            using (var pdbOutput = new MemoryStream())
+            {
+                var emitResult = comp2.Emit(output, pdbOutput, options: EmitOptions.Default.WithEmitMetadataOnly(true));
+                Assert.True(emitResult.Success);
+                emitResult.Diagnostics.Verify();
+                image2 = output.ToArray();
+            }
+
+            Assert.True(AssertEx.SequenceEqual(image1, image2) == expectMatch);
+        }
+
+        [Theory,
+            InlineData("public int M() { error(); }", true),
+            InlineData("public int M() { error() }", false), // Should be true. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
+            InlineData("public Error M() { return null; }", false), // Should be true. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
+            ]
+        public void RefAssemblyDiagnostics(string change, bool expectSuccess)
+        {
+            string sourceTemplate = @"
+public class C
+{
+    CHANGE
+}
+";
+            string source = sourceTemplate.Replace("CHANGE", change);
+            string name = GetUniqueName();
+            CSharpCompilation comp1 = CreateCompilationWithMscorlib(Parse(source),
+                options: TestOptions.DebugDll.WithDeterministic(true), assemblyName: name);
+
+            byte[] image;
+
+            using (var output = new MemoryStream())
+            using (var pdbOutput = new MemoryStream())
+            {
+                var emitResult = comp1.Emit(output, pdbOutput, options: EmitOptions.Default.WithEmitMetadataOnly(true));
+                Assert.Equal(expectSuccess, emitResult.Success);
+                Assert.Equal(!expectSuccess, emitResult.Diagnostics.Any());
+                image = output.ToArray();
+            }
+        }
+
+        [Fact]
+        public void VerifyRefAssembly()
+        {
+            string source = @"
+public class PublicClass
+{
+    public void PublicMethod() { }
+    private void PrivateMethod() { }
+    protected void ProtectedMethod() { }
+    internal void InternalMethod() { }
+}
+";
+            CSharpCompilation comp = CreateCompilation(source, references: new[] { MscorlibRef },
+                options: TestOptions.DebugDll.WithDeterministic(true));
+
+            var emitRefOnly = EmitOptions.Default.WithEmitMetadataOnly(true);
+
+            var verifier = CompileAndVerify(comp, emitOptions: emitRefOnly, verify: false); // PROTOTYPE(refout) PEVerify fails
```

Comment:
Tomasz and Vlad suggested emitting `throw null` instead of no method bodies, to make satisfy PEVerify. I'll look into that.